gh-91049: Introduce set vectorcall field API for PyFunctionObject

adphrost opened the pull request:
Avoid specializing functions with overridden vectorcall field

<!-- gh-issue-number: gh-91049 -->
* Issue: gh-91049
<!-- /gh-issue-number -->

Commit:
Enable setting vectorcall field on PyFunctionObjects

Commit:
check if vectorcall is nondefault before inlining call

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

itamaro commented on Lib/test/test_call.py:
```diff
@@ -665,6 +665,15 @@ def __call__(self, *args):
+        from _testcapi import pyobject_setvectorcall
+        def f(num): return num + 1
+        num = 10
+        assert f(num) is 11
```
```suggestion
        self.assertEqual(11, f(num))
```

itamaro commented on Lib/test/test_call.py:
```diff
@@ -665,6 +665,15 @@ def __call__(self, *args):
+        pyobject_setvectorcall(f)
+        # make sure specializer is triggered by running > 50 times
+        for _ in range(51):
+            assert f(num) is None
```
```suggestion
            self.assertIsNone(f(num))
```

itamaro commented on Python/ceval.c:
```diff
@@ -4792,7 +4792,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             int positional_args = total_args - KWNAMES_LEN();
             // Check if the call can be inlined or not
-            if (Py_TYPE(function) == &PyFunction_Type && tstate->interp->eval_frame == NULL) {
+            if (Py_TYPE(function) == &PyFunction_Type && tstate->interp->eval_frame == NULL && ((PyFunctionObject *)function)->vectorcall == _PyFunction_Vectorcall) {
```
this probably exceeds line length

adphrost replied:
What is the max line length allowed? And is there a linter I can run to handle this for me?

itamaro reviewed (commented):
thanks for working on this @adphrost !

vstinner commented on Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst:
```diff
@@ -0,0 +1,6 @@
+
+This PR introduces an API call where vectorcall field can be modified like so:
+
+`void PyFunction_SetVectorcall(PyFunctionObject *func, vectorcallfunc vectorcall);`
```
Please document the function in Doc/c-api/function.rst and in the NEWS entry, refer to use with:

```
:c:func:`PyFunction_SetVectorcall`
```

New public functions should also be documented in What's New in Python 3.11 > C API > New features: Doc/whatsnew/3.11.rst.

markshannon commented on Objects/funcobject.c:
```diff
@@ -208,6 +211,14 @@ PyFunction_SetDefaults(PyObject *op, PyObject *defaults)
     return 0;
 }
 
+int
```
This can't fail. Return `void`

markshannon commented on Modules/_testcapimodule.c:
```diff
@@ -6186,6 +6206,7 @@ static PyMethodDef TestMethods[] = {
     {"pyobject_fastcall", test_pyobject_fastcall, METH_VARARGS},
     {"pyobject_fastcalldict", test_pyobject_fastcalldict, METH_VARARGS},
     {"pyobject_vectorcall", test_pyobject_vectorcall, METH_VARARGS},
+    {"pyobject_setvectorcall", test_pyobject_setvectorcall, METH_VARARGS},
```
The names should match, and this only applies to functions. I'd suggest `function_setvectorcall`

markshannon commented on Modules/_testcapimodule.c:
```diff
@@ -5285,6 +5285,26 @@ test_pyobject_vectorcall(PyObject *self, PyObject *args)
+test_pyobject_setvectorcall(PyObject *self, PyObject *args)
+{
+    PyFunctionObject *func = NULL;
+    if (!PyArg_ParseTuple(args, "O", &func)) {
```
You can use `METH_O`. You'll also need to check that `func` is a function.

markshannon reviewed (commented):
A few comments.
Have you attempted to measure the performance impact of this? (I'd expect it to be small or negligible).

adphrost committed:
Apply suggestions from code review

Change how assertion is done

Co-authored-by: Itamar Ostricher <itamarost@gmail.com>

Commit:
addressed comments

Commit:
added to docs

Commit:
Merge branch 'main' into pyfunctionobject-set-vectorcall-field

Commit:
updated doc with fix by itamaro

Commit:
formatting

Commit:
removed doc from 3.11

Commit:
remove lines from 3.11

vstinner commented on Doc/c-api/function.rst:
```diff
@@ -83,6 +83,11 @@ There are a few functions specific to Python functions.
 
+.. c:function:: void PyFunction_SetVectorcall(PyFunctionObject *func, vectorcallfunc vectorcall)
+
+   Set the vectorcall field of a given function object *func*
```
```suggestion
   Set the vectorcall field of a given function object *func*.
```

Please add ``.. versionadded:: 3.12``.

vstinner commented on Modules/_testcapimodule.c:
```diff
@@ -5356,6 +5356,23 @@ test_pyobject_vectorcall(PyObject *self, PyObject *args)
+static PyObject *
+override_vectorcall(
+    PyObject *callable, PyObject *const *args, size_t nargsf, PyObject *kwnames) {
+    Py_RETURN_NONE;
```
Could you modify the test to return the string "overridden" instead of just returning None?

vstinner commented on Doc/whatsnew/3.12.rst:
```diff
@@ -236,6 +236,10 @@ New Features
   (Contributed by Wenzel Jakob in :gh:`93012`.)
 
+* Add new function :c:func:`PyFunction_SetVectorcall` to the C API
+  which sets the vectorcall field of a given PyFunctionObject
```
You may explain the purpose of the function here, like mention Cinder JIT and Pyjion.

markshannon replied:
I'd rather not have a "purpose". We should state clearly what the function does, not bless some particular use case.

@adphrost
Could you add a warning that the new function pointer must have **exactly** the same behavior as the unaltered function.
I am a little concerned that C extensions will abuse this, and we will have deal with the bug reports.

vstinner commented on Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst:
```diff
@@ -0,0 +1,6 @@
+CPython extensions providing optimized execution of Python bytecode (like Cinder JIT and Pyjion)
+can benefit from being able to modify the vectorcall field on instances of PyFunctionObject to allow calling the optimized path (e.g. JIT-compiled) directly.
+
+This PR introduces an API call where vectorcall field can be modified like so:
```
This text lands into https://docs.python.org/dev/whatsnew/changelog.html#changelog which is a changlog, you should not mention "a PR" there.

I suggest to just copy/paste what you wrote in whatsnew.

vstinner commented on Doc/whatsnew/3.12.rst:
```diff
@@ -236,6 +236,10 @@ New Features
   (Contributed by Wenzel Jakob in :gh:`93012`.)
 
+* Add new function :c:func:`PyFunction_SetVectorcall` to the C API
+  which sets the vectorcall field of a given PyFunctionObject
```
```suggestion
  which sets the vectorcall field of a given :c:type:`PyFunctionObject`.
```

vstinner reviewed (commented):
Oh. This change reminds me my old https://peps.python.org/pep-0510/ nice :-)

itamaro committed:
Apply review feedback from vstinner and markshannon

vstinner commented on Doc/c-api/function.rst:
```diff
@@ -83,6 +83,12 @@ There are a few functions specific to Python functions.
 
+.. c:function:: void PyFunction_SetVectorcall(PyFunctionObject *func, vectorcallfunc vectorcall)
+
+   Set the vectorcall field of a given function object *func*
```
```suggestion
   Set the vectorcall field of a given function object *func*.
```

vstinner commented on Doc/whatsnew/3.12.rst:
```diff
@@ -236,6 +236,12 @@ New Features
   (Contributed by Wenzel Jakob in :gh:`93012`.)
 
+* Add new function :c:func:`PyFunction_SetVectorcall` to the C API
+  which sets the vectorcall field of a given :c:type:`PyFunctionObject`
```
```suggestion
  which sets the vectorcall field of a given :c:type:`PyFunctionObject`.
```

vstinner commented on Doc/whatsnew/3.12.rst:
```diff
@@ -236,6 +236,12 @@ New Features
+* Add new function :c:func:`PyFunction_SetVectorcall` to the C API
+  which sets the vectorcall field of a given :c:type:`PyFunctionObject`
+  Warning: extensions using this API must preserve the behavior
+  of the unaltered function!
```
I don't think that this document is the right place to put a warning. Why this warning is not in the documentation of the function?

vstinner commented on Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst:
```diff
@@ -0,0 +1,5 @@
+which sets the vectorcall field of a given :c:type:`PyFunctionObject`
+
+Warning: extensions using this API must preserve the behavior
+of the unaltered function!
```
Please remove this warning.

vstinner commented on Misc/NEWS.d/next/C API/2022-05-03-19-35-37.gh-issue-92193.61VoFL.rst:
```diff
@@ -0,0 +1,5 @@
+Add new function :c:func:`PyFunction_SetVectorcall` to the C API
+which sets the vectorcall field of a given :c:type:`PyFunctionObject`
```
```suggestion
which sets the vectorcall field of a given :c:type:`PyFunctionObject`.
```

vstinner commented on Python/ceval.c:
```diff
@@ -4686,7 +4686,9 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
-            if (Py_TYPE(function) == &PyFunction_Type && tstate->interp->eval_frame == NULL) {
+            if (Py_TYPE(function) == &PyFunction_Type &&
+                tstate->interp->eval_frame == NULL &&
+                ((PyFunctionObject *)function)->vectorcall == _PyFunction_Vectorcall) {
```
I have a personal preference for this coding style:

```suggestion
            if (Py_TYPE(function) == &PyFunction_Type
                && tstate->interp->eval_frame == NULL
                && ((PyFunctionObject *)function)->vectorcall == _PyFunction_Vectorcall) {
```

Fidget-Spinner replied:
The author's code follows PEP 7 https://peps.python.org/pep-0007/#code-lay-out.

vstinner commented on Objects/funcobject.c:
```diff
@@ -134,6 +134,9 @@ uint32_t _PyFunction_GetVersionForCurrentState(PyFunctionObject *func)
         return func->func_version;
     }
+    if (func->vectorcall != _PyFunction_Vectorcall) {
+        return 0;
```
@markshannon @brandtbucher @gvanrossum: I don't know if this change is correct, I don't know how _PyFunction_GetVersionForCurrentState() is supposed to work.

markshannon replied:
This is correct. `_PyFunction_GetVersionForCurrentState` returns 0 if it cannot set the version number to a valid version.

itamaro committed:
Merge branch 'main' into gh-91049-set-vectorcall

itamaro committed:
Merge branch 'main' into gh-91049-set-vectorcall

itamaro committed:
Add deopt on func version in LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN

itamaro committed:
write func version to cache keys version when specializing LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN

itamaro committed:
remove redundant argcount check in LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN (replace with assert) - func version check is sufficient

itamaro committed:
Add missing periods in docs

itamaro committed:
move warning comment to the function C API docs

itamaro committed:
Merge branch 'main' into pyfunctionobject-set-vectorcall-field

itamaro committed:
fix race with GH-96519 (removed func_version in LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN specialization)

Fidget-Spinner reviewed (commented):
The parts touching the specializer and specialized opcodes LGTM.

markshannon commented on Python/ceval.c:
```diff
@@ -4138,7 +4141,9 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
-            if (Py_TYPE(function) == &PyFunction_Type && tstate->interp->eval_frame == NULL) {
+            if (Py_TYPE(function) == &PyFunction_Type &&
+                tstate->interp->eval_frame == NULL &&
+                ((PyFunctionObject *)function)->vectorcall == _PyFunction_Vectorcall) {
```
PEP 7 has the brace on the following line

itamaro replied:
fixed

markshannon reviewed (commented):
One minor formatting nit. Otherwise looks good.

itamaro committed:
PEP-7 formatting

erlend-aasland commented on Modules/_testcapi/vectorcall.c:
```diff
@@ -102,6 +102,23 @@ test_pyobject_vectorcall(PyObject *self, PyObject *args)
+function_setvectorcall(PyObject *self, PyObject *func)
+{
+    if (!PyFunction_Check(func)) {
+        PyErr_BadInternalCall();
```
This should be a TypeError.

```suggestion
        PyErr_SetString(PyExc_TypeError, "'func' must be a function");
```

erlend-aasland commented on Modules/_testcapi/vectorcall.c:
```diff
@@ -102,6 +102,23 @@ test_pyobject_vectorcall(PyObject *self, PyObject *args)
 
+static PyObject *
+override_vectorcall(
+    PyObject *callable, PyObject *const *args, size_t nargsf, PyObject *kwnames) {
```
PEP 7

```suggestion
override_vectorcall(PyObject *callable, PyObject *const *args, size_t nargsf,
                    PyObject *kwnames)
{
```

erlend-aasland commented on Objects/funcobject.c:
```diff
@@ -209,6 +212,13 @@ PyFunction_SetDefaults(PyObject *op, PyObject *defaults)
+void
+PyFunction_SetVectorcall(PyFunctionObject *func, vectorcallfunc vectorcall)
+{
+    func->func_version = 0;
```
Might consider adding an `assert(func != NULL)` here for debugging convenience.

```suggestion
    assert(func != NULL);
    func->func_version = 0;
```

erlend-aasland commented on Python/ceval.c:
```diff
@@ -3135,8 +3135,11 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             PyFunctionObject *f = (PyFunctionObject *)getattribute;
+            uint32_t func_version = read_u32(cache->keys_version);
+            assert(func_version != 0);
+            DEOPT_IF(f->func_version != func_version, LOAD_ATTR);
```
Is it possible to add a test for this deoptimisation?

itamaro replied:
is there an existing test I can use as a reference for how to trigger this specialization? I couldn't find any from a cursory search

Fidget-Spinner replied:
This specialisation triggers with the following code:

```python

class X:
    def __getattribute__(self, attr):
        return attr

x = X()
for _ in range(8):
    x.a  #LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN
```

itamaro replied:
thanks, I tried devising a test for this, PTAL

erlend-aasland replied:
Thanks! I gave it a try, but unfortunately, the test does not fail if I remove the deopt.

itamaro replied:
thanks for checking @erlend-aasland ! 

I realized I was testing something else (that specialization doesn't trigger for an already-overridden vectorcall)

pushed an update that now tests both scenarios (avoid specialization, and deopt) - verified the `test_setvectorcall_load_attr_specialization_deopt` fails when removing the deopt

erlend-aasland replied:
Thanks, looks good! :)

itamaro committed:
Address review feedback

- PEP-7 braces
- add assert
- raise TypeError

itamaro committed:
Add test for LOAD_ATTR specialization when overriding vectorcall of __getattribute__

Fidget-Spinner commented on Lib/test/test_call.py:
```diff
@@ -760,6 +760,32 @@ def __call__(self, *args):
+            self.assertEqual("a", x.a)
+        function_setvectorcall(X.__getattribute__)
+        # make sure specialization doesn't override the override
+        for _ in range(8):
```
Could you set a `QUICKENING_WARMUP_DELAY` constant and use it here please? We should try to unify these values in another PR. See for example https://github.com/python/cpython/blob/main/Lib/test/test_dis.py#L761.

itamaro replied:
done

erlend-aasland replied:
We should consider moving `QUICKENING_WARMUP_DELAY` to `test.support` in a follow-up PR.

Fidget-Spinner commented on Lib/test/test_call.py:
```diff
@@ -760,6 +760,32 @@ def __call__(self, *args):
+        x = X()
+        # trigger LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN specialization
+        for _ in range(8):
+            self.assertEqual("a", x.a)
```
Please pull both `self.assertEqual` out of the loops. This itself is a `LOAD_ATTR` and there have been situations in the past where it confounded some of the test results when the test fails.

so
```python
assertEqual = self.assertEqual
...
```

itamaro replied:
done

Fidget-Spinner commented on Lib/test/test_call.py:
```diff
@@ -760,6 +760,32 @@ def __call__(self, *args):
+        self.assertEqual(11, f(num))
+        function_setvectorcall(f)
+        # make sure specializer is triggered by running > 50 times
+        for _ in range(51):
```
Ditto about using `QUICKENING_WARMUP_DELAY` Maybe use `QUICKENING_WARMUP_DELAY * 5`

itamaro replied:
used `*10` to be sure :)

Fidget-Spinner reviewed (commented):
The test LGTM in general but I have a few comments. Thanks!

itamaro committed:
Improve setvectorcall + specialization testing

1. test specialization doesn't trigger when vectorcall is already overridden
2. test specialization gets deopted when vectorcall is overridden after specialization triggered

markshannon reviewed (approved):
Looks good now.

itamaro committed:
Merge branch 'main' into pyfunctionobject-set-vectorcall-field
